diff --git a/src/udisksdaemon.c b/src/udisksdaemon.c
index 51e2bb5c..f5d594a0 100644
--- a/src/udisksdaemon.c
+++ b/src/udisksdaemon.c
@@ -397,12 +397,17 @@ udisks_daemon_constructed (GObject *object)
         }
     }
 
-  if (!g_file_test (PACKAGE_LOCALSTATE_DIR "/lib/udisks2", G_FILE_TEST_IS_DIR))
+  /* If we're running in a snap environment just keep the state in $SNAP_DATA.
+   * otherwise we need to create a traditional state dir */
+  if (g_getenv ("SNAP_DATA") == NULL)
     {
-      if (g_mkdir_with_parents (PACKAGE_LOCALSTATE_DIR "/lib/udisks2", 0700) != 0)
-        {
-          udisks_critical ("Error creating directory %s: %m", PACKAGE_LOCALSTATE_DIR "/lib/udisks2");
-        }
+      if (!g_file_test (PACKAGE_LOCALSTATE_DIR "/lib/udisks2", G_FILE_TEST_IS_DIR))
+      {
+        if (g_mkdir_with_parents (PACKAGE_LOCALSTATE_DIR "/lib/udisks2", 0700) != 0)
+          {
+            udisks_error ("Error creating directory %s: %m", PACKAGE_LOCALSTATE_DIR "/lib/udisks2");
+          }
+      }
     }
 
 
diff --git a/src/udisksstate.c b/src/udisksstate.c
index 0dff3230..438e83fd 100644
--- a/src/udisksstate.c
+++ b/src/udisksstate.c
@@ -2374,6 +2374,9 @@ udisks_state_get_modules (UDisksState *state)
 static gchar *
 get_state_file_path (const gchar *key)
 {
+  if (g_getenv ("SNAP_DATA") != NULL)
+    return g_strdup_printf ("%s/%s", g_getenv ("SNAP_DATA"), key);
+
   if (g_str_equal (key, UDISKS_STATE_FILE_MOUNTED_FS_PERSISTENT))
     return g_strdup_printf (PACKAGE_LOCALSTATE_DIR "/lib/udisks2/%s", key);
 
